### ISSUE TITLE
[super-parameters] Lint positional parameter name != name of constructor parameter

### ISSUE THREAD

**Issue by eernstg** [dart-lang/linter#3107]
Consider the following library, which is using the upcoming `super-parameters` feature:

```dart
class A {
  final int x;
  A(this.x);
}

class B extends A {
  B(super.x); // <-- Note the name `x`.
}
```

The `super.x` parameter declaration is correct according to the specification of the new feature, but it has a peculiar property: We could change the name to _anything_ we want, and it does not matter!

The reason for this is that the chosen name will be used when the desugaring step that defines the semantics of super parameters is performed, and for a positional super parameter any name will work (as long as we avoid name clashes).

**This issue is a proposal to create a lint that flags a positional super parameter _p_ whose name differs from the name of the associated positional constructor parameter.**

This means that the above example is fine, but `B(super.id)` would be flagged for any identifier `id` which isn't `x`.

It is still somewhat weird that the name of a positional super parameter is declared and (typically) never used, but with this lint in place it is at least possible to rely on the name to say something about the purpose and meaning of the parameter, insofar as the given super-constructor parameter does the same.

**Comment by pq** [comment]
This sounds like a great idea.   Thematically it's a bit like [`avoid_renaming_method_parameters`](https://dart-lang.github.io/linter/lints/avoid_renaming_method_parameters.html) so maybe `avoid_renaming_super_parameters`?  

**Bike shed alert:** we've gradually veered away from names that correspond to style guide prescriptive wording (which is to say avoiding "avoid" and "prefer") so we might consider alternative names.  It's a trade-off though...

**Comment by bwilkerson** [comment]
> ... for a positional super parameter any name will work ...

I haven't yet read the proposal for this feature, but does that mean that positional parameters are passed in the order in which they are declared? For example, given the following
```dart
class A {
  final int x, y;
  A(this.x, this.y);
}
class B extends A {
  B(super.y, super.x);
}
void main() {
  print(B(1, 2).x);
}
```
will it print `1`? Because if that's the case, then it could be _very_ confusing to users to not have the names match. I have to wonder whether a lint is sufficient, though it's probably fine as long as it's in the recommended set.

> ... we've gradually veered away from names that correspond to style guide prescriptive wording ...

That's true, but I think we shouldn't refuse to use 'avoid' when it isn't expressing a prescriptive label from the style guide, as in this case (although I could see other names such as `no_renamed_super_parameter`). It's a perfectly good word, the only problem with using it is when the style guide changes because it's hard to change the name of a lint. ('prefer' is less useful for lint names in general, and so easier to just refuse to consider).

**Comment by eernstg** [comment]
> will it print `1`?

Yes, the positional formal super parameter number _k_ is passed as the positional actual argument number _k_ in the superinitializer (and if there is one or more positional super parameters then there can't be any positional parameters that aren't super parameters, so there's never a conflict about position number _k_).

I agree that reordering the names would be _very_ confusing, and using entirely different names isn't helpful either. Surely it could be a hint. If a developer seriously wants to do some renaming then it is always possible to specify all the positional parameters directly.
